### PR TITLE
Async API for GAPI

### PR DESCRIPTION
**Pull request by anton-potapov**
- naive implementation as a starting point

### This pullrequest changes
- Adds a naive implementation of Asynchronous API for GAPI module

### CODE REVIEW

File: modules/gapi/CMakeLists.txt [hunk 1]
```diff
@@ -61,6 +61,9 @@ set(gapi_srcs
     # Executor
     src/executor/gexecutor.cpp
 
+    #Async
+    src/compiler/gasync.cpp
```
Comment (dmatveev) [inline modules/gapi/CMakeLists.txt]: Why it is in `src/compiler` by the way? We have more relevant `src/executor` instead.

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018 Intel Corporation
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: 2019

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+#include <exception>        //for std::exception_ptr
+#include <functional>       //for std::function
+#include "opencv2/gapi/garg.hpp"
+
+namespace cv {
+    //fwd declaration
+    class GCompiled;
+
+    GAPI_EXPORTS void                async(GCompiled& gcmpld, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs);
+    GAPI_EXPORTS std::future<void>   async(GCompiled& gcmpld, GRunArgs &&ins, GRunArgsP &&outs);
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: `namespace gapi`?

`cv::async` may look too general for OpenCV

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp (hunk 2 above)
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: (everywhere)

File: modules/gapi/include/opencv2/gapi/gcomputation_async.hpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+#include "opencv2/gapi/garg.hpp"            //for GRunArgs, GRunArgsP
+#include "opencv2/gapi/gcommon.hpp"         //for GCompileArgs
+
+namespace cv {
+    //fwd declaration
+    class GComputation;
+
+    GAPI_EXPORTS void                async_apply(GComputation& gcomp, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs, GCompileArgs &&args = {});
+    GAPI_EXPORTS std::future<void>   async_apply(GComputation& gcomp, GRunArgs &&ins, GRunArgsP &&outs, GCompileArgs &&args = {});
+}
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/gcomputation_async.hpp]: also `namespace gapi`?

File: modules/gapi/src/compiler/gasync.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        auto f = prms.value.get_future();
+        auto l = [=]() mutable {
+            gcmpld(std::move(ins), std::move(outs));
+            prms.value.set_value();
+        };
+
+        impl::the_ctx.add_task(l);
+        return f;
+    }
+}
```
Comment (dmatveev) [inline modules/gapi/src/compiler/gasync.cpp]: Is there anywhere any info describing this?

File: modules/gapi/src/compiler/gasync.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+#include <condition_variable>
+
+#include <future>
+#include <condition_variable>
+//#include <chrono>
+#include <stdexcept>
+#include <queue>
+
+namespace {
+    //This is a tool to to move initialize captures of a lambda in C++11
```
Comment (dmatveev) [inline modules/gapi/src/compiler/gasync.cpp]: to to

File: modules/gapi/src/compiler/gasync.cpp (hunk 5 above)
Comment (dmatveev) [reply modules/gapi/src/compiler/gasync.cpp]: (seems I got it after reading, but anyway overall description that "now it is just a thread with a queue, calling gcompileds followed by their completion callbacks" is kinda missing

**Comment by dmatveev** [comment]
Seems ok, fix minor stuff and add missing pieces (and make build green of course).

Thanks

File: modules/gapi/test/gapi_async_test.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    std::future<void> f;
+
+    template<typename... Args >
+    void start_async(Args&&... args){
+        f = this->crtp_cast_(this)->async(std::forward<Args>(args)...);
+    }
+
+    void wait_for_result()
+    {
+        //FIXME: add a reasonable timeout here ?
```
Comment (SDxKeeper) [inline modules/gapi/test/gapi_async_test.cpp]: Can you add timeout argument to the function, with default value (1 min?)

**Comment by anton-potapov** [comment]
@dmatveev @rgarnov  @dbudniko please take a look

File: modules/gapi/src/compiler/gasync.cpp [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+       move_through_copy(T&& g) : value(std::move(g)) {}
+       move_through_copy(move_through_copy&&) = default;
+       move_through_copy(move_through_copy const& lhs) : move_through_copy(std::move(const_cast<move_through_copy&>(lhs))) {}
+    };
+}
+
+namespace cv {
+
+    namespace impl{
+        class async_service {
```
Comment (alalek) [inline modules/gapi/src/compiler/gasync.cpp]: Please avoid indentation in namespaces.

File: modules/gapi/src/compiler/gasync.cpp [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                }
+            }
+            ~async_service(){
+                if (thrd.joinable())
+                {
+                    exiting = true;
+                    mtx.lock();
+                    mtx.unlock();
+                    cv.notify_one();
+                    thrd.join();
```
Comment (alalek) [inline modules/gapi/src/compiler/gasync.cpp]: Be careful with waiting for notifications from worker threads during "cleanup".
Win32 `ExitProcess()` terminates all non-main threads **unconditionally** - probably with broken synchronization state - holding mutexes (user space at least) and other synchronization primitives.
And after that, it calls `atexit()` handlers from shared libraries, like this destructor.

Hopefully `thrd.joinable()` can handle that properly.

File: modules/gapi/src/compiler/gasync.cpp [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+                }
+            }
+        };
+
+        async_service the_ctx;
```
Comment (alalek) [inline modules/gapi/src/compiler/gasync.cpp]: Probably `static` (or anonymous namespace)

What is about initialization "on-demand"?

static
async_service& getAsyncService()
{
    static async_service g_asyncService;
    return g_asyncService;
}
...
the_ctx. => getAsyncService().
```

File: modules/gapi/include/opencv2/gapi/gcomputation_async.hpp (hunk 4 above)
Comment (anton-potapov) [reply modules/gapi/include/opencv2/gapi/gcomputation_async.hpp]: the problem this this approach is that ADL will not work then (as ```async``` and ```GCompiled```) will be in different namespaces. i.e. following code will not compile without either full qualification of the function (```cv::gapi::async_apply```) or ```using```  directive/clause: 

```c++
cv::GMat in;  
cv::GMat  out = in + in;
cv::GCompuation g{in,out};
cv::Mat m, r;
//here will be a compile error
// async_apply(g, cv::gin(m), cv::gout(r));

// use it either like this:
cv::gapi::async_apply(g, cv::gin(m), cv::gout(r));

// or like this:
using cv::gapi::async_apply;

async_apply (g, cv::gin(m), cv::gout(r));
```

maybe add **g** to name - i.e. ```gasync_apply```?

File: modules/gapi/CMakeLists.txt (hunk 1 above)
Comment (anton-potapov) [reply modules/gapi/CMakeLists.txt]: fixed

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp (hunk 2 above)
Comment (anton-potapov) [reply modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: fixed

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp (hunk 3 above)
Comment (anton-potapov) [reply modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: moved to ```gapi::wip``` namespace

File: modules/gapi/include/opencv2/gapi/gcomputation_async.hpp (hunk 4 above)
Comment (anton-potapov) [reply modules/gapi/include/opencv2/gapi/gcomputation_async.hpp]: fixed

File: modules/gapi/src/compiler/gasync.cpp (hunk 5 above)
Comment (anton-potapov) [reply modules/gapi/src/compiler/gasync.cpp]: added

File: modules/gapi/src/compiler/gasync.cpp (hunk 6 above)
Comment (anton-potapov) [reply modules/gapi/src/compiler/gasync.cpp]: thanks - fixed

File: modules/gapi/test/gapi_async_test.cpp (hunk 7 above)
Comment (anton-potapov) [reply modules/gapi/test/gapi_async_test.cpp]: After some thinking it out - it was a bad idea to add timeout - as there is no way to cancel the work gracefully , so timeout simply meaningless/confusing here, so TODO is removed

File: modules/gapi/test/gapi_async_test.cpp [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+template<typename computation_t>
+using cases = ::testing::Types<
+            Case<computation_t, CallBack, AsyncCompiled>,
+            Case<computation_t, CallBack, AsyncApply>,
+            Case<computation_t, Future,   AsyncCompiled>,
+            Case<computation_t, Future,   AsyncApply>
+            >;
+INSTANTIATE_TYPED_TEST_CASE_P(AsyncAPINormalFlow_,        normal,     cases<SumOfSum2x2>);
+INSTANTIATE_TYPED_TEST_CASE_P(AsyncAPIExecptionHandling_, exception,  cases<ExceptionOnExecution>);
```
Comment (alalek) [inline modules/gapi/test/gapi_async_test.cpp]: typo in `Exception`

Comment (anton-potapov) [reply modules/gapi/test/gapi_async_test.cpp]: thanks - fixed

File: modules/gapi/src/compiler/gasync.cpp (hunk 10 above)
Comment (dmatveev) [reply modules/gapi/src/compiler/gasync.cpp]: +1 for on-demand

File: modules/gapi/test/gapi_async_test.cpp [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+        return GComputation{in, out};
+    });
+
+    const cv::Size sz{2, 2};
+    cv::Mat in_mat{sz, CV_8U, cv::Scalar(1)};
+    cv::Mat out;
+
+    auto f = cv::gapi::wip::async_apply(self_mul,cv::gin(in_mat), cv::gout(out));
+    f.wait();
+}
```
Comment (dmatveev) [inline modules/gapi/test/gapi_async_test.cpp]: Why it is DISABLED BTW?

File: modules/gapi/src/executor/gasync.cpp [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+                    mtx.lock();
+                    mtx.unlock();
+                    cv.notify_one();
+                    thrd.join();
+                }
+            }
+        };
+
+        async_service the_ctx;
+    }
```
Comment (dmatveev) [inline modules/gapi/src/executor/gasync.cpp]: This whole block of code still lacks comments. "What author has wanted to say?"

File: modules/gapi/src/compiler/gasync.cpp (hunk 8 above)
Comment (anton-potapov) [reply modules/gapi/src/compiler/gasync.cpp]: fixed

File: modules/gapi/src/executor/gasync.cpp (hunk 13 above)
Comment (anton-potapov) [reply modules/gapi/src/executor/gasync.cpp]: see below at [line 128](https://github.com/opencv/opencv/pull/14346/files/93213d9b36121c61cec7b4a1297bd437125d5f00#diff-a6eac27ed59a18bd5f4bebfe2a80efe5R128)

Is it enough - or more comments are required ?

File: modules/gapi/src/compiler/gasync.cpp (hunk 10 above)
Comment (anton-potapov) [reply modules/gapi/src/compiler/gasync.cpp]: Thread It is already on created on - demand - during first call of ```add_task()``` [async.cpp line 49](https://github.com/opencv/opencv/pull/14346/files/93213d9b36121c61cec7b4a1297bd437125d5f00#diff-a6eac27ed59a18bd5f4bebfe2a80efe5R49)

File: modules/gapi/test/gapi_async_test.cpp (hunk 12 above)
Comment (anton-potapov) [reply modules/gapi/test/gapi_async_test.cpp]: enabled back

File: modules/gapi/src/executor/gasync.cpp (hunk 13 above)
Comment (anton-potapov) [reply modules/gapi/src/executor/gasync.cpp]: fixed the code and added comments

**Comment by anton-potapov** [comment]
@dbudniko  please review , it seems that @dmatveev  is not reachable ... 
@alalek could you please change the reviewer to @dbudniko  or review it yourself ? (All design issues were resolved, only fixes for technical issues are left for review )

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019 Intel Corporation
+
+
+#ifndef OPENCV_GAPI_GCOMPILED_ASYNC_HPP
+#define OPENCV_GAPI_GCOMPILED_ASYNC_HPP
+
+#include <future>
```
Comment (dbudniko) [inline modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: //for std::future 
to align with next std headers comments.

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+#include <exception>        //for std::exception_ptr
+#include <functional>       //for std::function
+#include "opencv2/gapi/garg.hpp"
+
+namespace cv {
+    //fwd declaration
+    class GCompiled;
+
+namespace gapi{
+namespace wip {
```
Comment (dbudniko) [inline modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: May be it worth to choose better name for namespace? Something containing "volatile".

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    GAPI_EXPORTS void                async(GCompiled& gcmpld, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs);
```
Comment (dbudniko) [inline modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: Why do we need these overloads? Please provide comments for usage scenario for both.

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    GAPI_EXPORTS std::future<void>   async(GCompiled& gcmpld, GRunArgs &&ins, GRunArgsP &&outs);
+} // namespace gapi
+} // namespace expreimental
```
Comment (dbudniko) [inline modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: Should be wip?

File: modules/gapi/include/opencv2/gapi/gcomputation_async.hpp [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+#include <functional>                       //for std::function
+#include "opencv2/gapi/garg.hpp"            //for GRunArgs, GRunArgsP
+#include "opencv2/gapi/gcommon.hpp"         //for GCompileArgs
+
+namespace cv {
+    //fwd declaration
+    class GComputation;
+namespace gapi {
+namespace wip  {
+    GAPI_EXPORTS void                async_apply(GComputation& gcomp, std::function<void(std::exception_ptr)>&& callback, GRunArgs &&ins, GRunArgsP &&outs, GCompileArgs &&args = {});
```
Comment (dbudniko) [inline modules/gapi/include/opencv2/gapi/gcomputation_async.hpp]: Again, please, provide comments regarding these overloads.

File: modules/gapi/src/executor/gasync.cpp [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+#include "opencv2/gapi/gcomputation_async.hpp"
+#include "opencv2/gapi/gcomputation.hpp"
+#include "opencv2/gapi/gcompiled_async.hpp"
+#include "opencv2/gapi/gcompiled.hpp"
+
+#include <condition_variable>
+
+#include <future>
+#include <condition_variable>
+//#include <chrono>
```
Comment (dbudniko) [inline modules/gapi/src/executor/gasync.cpp]: Please remove

File: modules/gapi/src/executor/gasync.cpp (hunk 13 above)
Comment (dbudniko) [reply modules/gapi/src/executor/gasync.cpp]: I still don't understand what is happening here. Queue, second queue (local instance). What is the purpose of this async_service? Why it's so complicated?

File: modules/gapi/src/executor/gasync.cpp [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+        p.set_exception(eptr);
+    }
+    else {
+        p.set_value();
+    }
+}
+}//namespace
+
+//For now these async functions are simply wrapping serial version of apply/operator() into a functor.
+//These functors are then serialized into single queue, which is when processed by a devoted background thread.
```
Comment (dbudniko) [inline modules/gapi/src/executor/gasync.cpp]: "when" - "then" missprint?

File: modules/gapi/test/gapi_async_test.cpp (hunk 12 above)
Comment (dbudniko) [reply modules/gapi/test/gapi_async_test.cpp]: No comments at all. What is the general idea behind these tests?

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp (hunk 15 above)
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: No, `wip` was my proposal (communicated to Anton "verbally")

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp (hunk 17 above)
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: Maybe `experimental` is also a good choice, we just need to use one-for-all

File: modules/gapi/src/executor/gasync.cpp [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+                        if (q.empty())
+                        {
+                            //block current thread until arrival of exit request or new elements
+                            cv.wait(lck, [&](){ return exiting || !q.empty();});
+                        }
+                        //usually swap for std::queue is plain pointers exchange, so relatively cheap
+                        q.swap(second_q);
+                        lck.unlock();
+
+                        while (!second_q.empty())
```
Comment (dmatveev) [inline modules/gapi/src/executor/gasync.cpp]: should this `while` condition be extended with `exiting` check as well?

What is the policy on `exiting`? Should be wait for all issues request or just "forget" it?

File: modules/gapi/src/executor/gasync.cpp (hunk 13 above)
Comment (dmatveev) [reply modules/gapi/src/executor/gasync.cpp]: Is there any reply on @alalek's concerns regarding https://github.com/opencv/opencv/pull/14346#discussion_r276182722 ?

File: modules/gapi/include/opencv2/gapi/gcompiled_async.hpp (hunk 17 above)
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/gcompiled_async.hpp]: Fixed

File: modules/gapi/src/executor/gasync.cpp (hunk 20 above)
Comment (anton-potapov) [reply modules/gapi/src/executor/gasync.cpp]: thanks, removed

**Comment by dmatveev** [comment]
👍  @alalek  can you please proceed with merge?

**Pull request merged**